lib/src/revset.rs

```diff
@@ -322,6 +323,25 @@ pub enum RevsetCommitRef {
     GitHead,
 }
 
+/// A custom revset filter expression, defined by an extension.
+pub trait RevsetFilterExtension: std::fmt::Debug + Any {
+    fn as_any(&self) -> &dyn Any;
+
+    /// Returns true iff this filter matches the specified commit.
+    fn matches_commit(&self, commit: &Commit) -> bool;
+
+    /// Returns true iff this filter is functionally equivalent to `ext`.
+    fn eq_ext(&self, ext: &dyn RevsetFilterExtension) -> bool;
```

Comment:
Perhaps, we can remove `Eq, PartialEq` from `Expression` enums. I don't think all future core filter expressions do support Eq.

I'll take a look.

Reply:
Hmm, it seems better to rewrite the parser first so that parsing tests don't depend on equivalence of `RevsetExpression` trees. Until then, maybe we can use pointer-based equivalence? I assume the backend doesn't need `eq_ext()`, and will use `as_any()` to extract the underlying representation.

https://doc.rust-lang.org/std/rc/struct.Rc.html#method.ptr_eq